### DISCUSSION TITLE
Different electricity prices depending on time and day of the week

### DISCUSSION THREAD

**Discussion by KnightwhosayNi777** [teslamate-org/teslamate#772] [Ideas]
**What would you like to be added**: Different electricity prices depending on time and day of the week

**Why is this needed**: Our electricity supplier uses different electricity tariffs depending on the time of day. 

**Additional context**:

**Comment by DrMichael** [comment]
I get more and more the feeling, that that should be covered through a different app.
Teslamate could provide date and time, duration, location and kWh as an export...

**Comment by Dulanic** [comment]
Yeah.... I have the same type of electricity and I just ALWAYS charge at night when it's cheapest. You can always override the cost for those one off situations.  That's a lot of time spent for a tiny use. Wouldn't you always use a set time to start charging? Thats what I do... it doesn't start charging till midnight.

**Comment by petariga** [comment]
It is a nice idea, however for example in our area the timing for different rates is very flexible and changes everyday, so you can't really make some schedule or system which you could put into TeslaMate...

**Comment by baylanger** [comment]
I don't have time this week but I'd like to propose writing a small script that would run via cron.

For the docker image, we add a "cron" pkg.  In the documentations show 1-2 script examples for cases that we know users have asked for and show how to enable their script.

I would go with a bash script but I haven't check all the possibilities that Postgres offers, I'd think for many cases perhaps Postgres would be better.  Comments are welcome :) 

I'll tell you another charge cost case that I'm well aware.  On each day, the cost of the first 40kWh is ~20-25% than when reaching 40kWh.  In this case, a device over mqtt could perhaps report the current total kWh since midnight - otherwise something over http.  This isn't so simple to implement and I'm positive some electric companies likely have even more odd cases.  I know many solar panel owners have devices that work with mqtt, this seems another difficult case for teslamate to calculate real cost... if the sun is out, there's power coming from the panels otherwise it might all be from the electric company or a powerwall?

**Comment by tacotran** [comment]
Whatever solution is discovered here, it will very likely have great use over in https://github.com/adriankumpf/teslamate/issues/655

Or vice-versa.

**Comment by paulobrien** [comment]
We have cheap electricity at night between set hours, so this would be very useful.

**Comment by Dulanic** [comment]
@paulobrien I do also so the easy work around is set your kwh price to your night price. I never charge during peak times, do you?

**Comment by ronnyandre** [comment]
This feature would be awesome. My electricity provider has an excellent API that gives me realtime electricity prices, which would be awesome to implement so I can have real actual charge pricing at home.

**Comment by tobiasehlert** [comment]
@KnightwhosayNi777 @Dulanic @paulobrien @ronnyandre 

I think you guys should check out TeslaMateAgile project, since it might just do what you are looking for. :)

https://github.com/MattJeanes/TeslaMateAgile

Regards,
Tobias

**Reply by baylanger** [reply]
@tobiasehlert ~~ain’t that site only for “octopusenergy” clients?~~ needed more coffee.  I had check that project before and it was not good option because it was supporting only that company.  I see commits added following providers :

```
Tibber (nov 2020)
Fixed Price (manually specify prices for different times of the day) (feb 4 2021)
```

That project can be used if the provider is supported ... if it was written in Elixir or Python, might be easier to add to TeslaMate instead of running another process that requires C#.  From my knowledge, there are still plenty of cases missing.

- prices based on daily consumption, `if total_kW_consumed_today > x kW => price_increase()` - counter resets each day midnight.  Requirements : "live consumption updated every ~15min max" or install and use a smart device on the main electric panel with update to the ~ minute.
- prices based on outdoor temperature, `if outdoor_temp < xx => price_increase else default_price()` - use if possible a worldwide weather online provider OR install and use a smart device/weather station.
- possibly other cases above^

TeslaMate already talks mqtt for energy battery systems/solar/wind, etc.  I'm not sure what's possible yet and not but such end users would probably want something like :

- is the main grid provider down?  if so, you might prefer to keep your remaining battery storage for your house and not the car.
- if your "house-grid" generates 3kW and car pulls 6kW, the cost from the utility is "total kWh to car / 2".   But solar/wind kW goes up/down non-stop, 

Based on the above, I hope someday TeslaMate might have logic based on "electric metrics reading (consumption vs generated by house-grid" that could start/stop charging based.  Yes I plan to have solar and/or wind, battery, etc!  Adding the start/stop logic based on metrics providers (mqtt, weather, daily consumption, etc) is very simple, writing the actual "metrics providers" that's another story.

**Reply by tobiasehlert** [reply]
No @baylanger, the functionality has been extended to support other charge-calculations as well :-)

**Reply by baylanger** [reply]
Per your reply, you might not have noticed the "edit" of my answer.  But correct, more providers have now been added.  @tobiasehlert

**Reply by tobiasehlert** [reply]
Sorry @baylanger, the edit did not come with the email notification, so missed your update. =D

You actually got a point there as well with the solar generations for example. I do have solar (in around 6-7 weeks), so the cost would not be reflected correctly in this case (if I would charge at night). Since I use Tibber myself, I only charge at nights when the prices are low anyways, so I would probably not face that situation right now.

And the approach with amount charged is actually different from the chargers perspective and TeslaMate as well, since I do not run SentryMode at home, the car sleep a bit longer so that TeslaMate doesn't record the first couple of kWh.. the chargebox data and TeslaMate is differing around 10% which is kind of interesting in my eyes. With SentryMode on, it's not so much difference, but still some difference :)

**Reply by baylanger** [reply]
@tobiasehlert I'm not sure 100% what you mean by the `charged is actually different` but I'm wondering if you're not referring to this that I had open,

https://github.com/adriankumpf/teslamate/issues/748

Discussion how it's calculate is provided in there.

**Reply by baylanger** [reply]
the above references also https://github.com/adriankumpf/teslamate/issues/290

**Reply by tobiasehlert** [reply]
@baylanger well, probably not the same as you mentioned above.

I've noticed that when my charger at home kicks off charging, the car is not always walking up TeslaMate so quick, so it's charging without getting it logged. But when SentryMode is enabled, it's much quicker. =P
But yeah.. this is not related to this discussion in the first place. Was more of a side note only hehe
